### PR TITLE
Don't add revnumber attr when project.version is unspecified

### LINKED ISSUE #329: Do not add revnumber if no project.version
If project.version is not defined then `AbstractAsciidoctorTask.getTaskSpecificDefaultAttributes` should not add `revnumber` as an attribute.

### PR DESCRIPTION
**Pull request by lhotari**
- Fixes #329
- project.version defaults to 'unspecified' in Gradle
- fix attempt was previously made in https://github.com/asciidoctor/asciidoctor-gradle-plugin/pull/337/commits/eb75b4972eb50e9c29e6a97e34c1b75a58bed95c, but that didn't check for `Project.DEFAULT_VERSION` (which is `unspecified`).

### CODE REVIEW

File: base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy [hunk 1]
```diff
@@ -770,7 +771,7 @@ abstract class AbstractAsciidoctorBaseTask extends DefaultTask {
             'gradle-project-name': (Object) project.name
         ]
 
-        if (project.version != null) {
+        if (project.version != null && project.version.toString() != Project.DEFAULT_VERSION) {
```
Comment (ysb33r) [inline base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy]: EMI, but what does `project.DEFAULT_VERSION` provide ?

**Comment by ysb33r** [comment]
Hey @lhotari Can you base your PR against `development-3.x`, please?

File: base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy (hunk 1 above)
Comment (lhotari) [reply base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy]: "unspecified" https://github.com/gradle/gradle/blob/7095e7c2c00f1d10683596f8beba9ca9935c04a0/subprojects/core-api/src/main/java/org/gradle/api/Project.java#L238

**Comment by lhotari** [comment]
> Hey @lhotari Can you base your PR against `development-3.x`, please?

Yes. working on it.

**Comment by lhotari** [comment]
It's rebased now. I also added another fix in this same PR that I came across at the same time. I can move that to a separate PR if that's required. The 2nd commit doesn't add `revnumber@` when it has been specified by the user.

**Comment by ysb33r** [comment]
>  I can move that to a separate PR if that's required. 

Keep in the same one. There is enough material being contributed such that we can release 3.1.0 shortly. // cc @Mogztter

**Comment by ysb33r** [comment]
@lhotari There still seem to be one test failure.

**Comment by lhotari** [comment]
> There still seem to be one test failure.

will check it now.

**Comment by lhotari** [comment]
@ysb33r I pushed a fix for the test failure, https://github.com/asciidoctor/asciidoctor-gradle-plugin/pull/512/files#diff-5479704be335de86c1fe2110ef764cbbR1033-R1040 . Had to use CompileDynamic to resolve the issue. It seems that the signature of `.collect` changed in Groovy 2.5.0 .

File: base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy [hunk 2]
```diff
@@ -1029,6 +1030,14 @@ abstract class AbstractAsciidoctorBaseTask extends DefaultTask {
         defaultAttrs
     }
 
+    // must use dynamic invocation to invoke collect method for compatibility with Groovy < 2.5.0
+    @CompileDynamic
+    private Set<String> trimOverridableAttributeNotation(Set<String> attributeKeys) {
+        // remove possible trailing '@' character that is used to encode that the attribute can be overridden
+        // in the document itself
+        attributeKeys.collect { k -> k - ~/@$/ } as Set
```
Comment (ysb33r) [inline base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy]: We have a utility to deal with that which will avoid dynamic compilation.
```groovy
Transform.toSet (attributeKeys) { k -> k - ~/@$/ }
```

Comment (lhotari) [reply base/src/main/groovy/org/asciidoctor/gradle/base/AbstractAsciidoctorBaseTask.groovy]: Thanks. good to know. Using `Transform.toSet` now.

**Pull request merged**